File: frontend/pifitness/src/lib/types/volleyball.ts
/**
 * Volleyball Scorekeeping Types
 *
 * Cross-surface data contract for the Beach Volleyball scorekeeping feature.
 * Shared between frontend (React) and backend (FastAPI Pydantic models).
 *
 * Maps directly to the pre-built backend tables (no migrations; designs 006-001 + 006-002):
 *   volleyball.games   (game lifecycle; Team A is always "SR"; 006-002 adds
 *                       partner_number/partner_name for match context)
 *   volleyball.points  (one row per scored point — event sourcing;
 *                       score is derived by counting rows, never stored;
 *                       006-002 adds event_type for notable-play tags)
 *
 * Endpoints (backend/api/volleyball.py, prefix /api/sports/volleyball):
 *   GET    /api/sports/volleyball                      -> game history (completed_at desc, w/ final scores)
 *   GET    /api/sports/volleyball/active               -> active game + points + derived score, or game: null
 *   POST   /api/sports/volleyball                      -> create game (opponent name + SR partner); 409 if one is active
 *   POST   /api/sports/volleyball/{id}/points          -> add a point for a team
 *   DELETE /api/sports/volleyball/{id}/points/{team}   -> remove that team's most recent point
 *   POST   /api/sports/volleyball/{id}/points/latest/event -> tag the most recent point (either team) with an event_type (006-002)
 *   POST   /api/sports/volleyball/{id}/end             -> end game (completed_at = MAX(recorded_at))
 *   DELETE /api/sports/volleyball/{id}                 -> abandon game (cascade deletes its points)
 */

/** Lifecycle status of a volleyball game. */
export type VolleyballGameStatus = 'active' | 'completed';

/** Scoring team discriminator (CHECK in volleyball.points). */
export type VolleyballScoringTeam = 'SR' | 'OPPONENT';

/** Notable-play event types taggable on a point (006-002). */
export type VolleyballEventType = 'Ace' | 'Block' | 'Spike' | 'Dive';

/**
 * A volleyball game.
 * Maps to volleyball.games. `started_at` is set by the app to
 * MIN(recorded_at) of the game's points (null until the first point);
 * `completed_at` to MAX(recorded_at) when ended — never button-press time.
 */
export interface VolleyballGame {
  /** Primary key (SERIAL). */
  game_id: number;
  /** Opponent team name (Team B). Team A is always "SR" and is not stored. */
  team_b_name: string;
  /** Lifecycle status. */
  status: VolleyballGameStatus;
  /** MIN(recorded_at) of this game's points, or null before the first point. */
  started_at: string | null;
  /** MAX(recorded_at) of this game's points once completed. */
  completed_at: string | null;
  /** SR side's partner jersey number for this match (006-002). Nullable in the
   *  DB for pre-006-002 rows; required by the API at creation going forward. */
  partner_number: number | null;
  /** SR side's partner name for this match (006-002; optional). */
  partner_name: string | null;
  /** Optional label (unused in v1). */
  label?: string | null;
  /** Optional notes (unused in v1). */
  notes?: string | null;
  /** Row creation time. */
  created_at: string;
}

/**
 * A single scored point.
 * Maps to volleyball.points. The running score is derived by counting
 * these rows per team (event sourcing) — no score value is ever stored.
 */
export interface VolleyballPoint {
  /** Primary key (SERIAL). */
  point_id: number;
  /** Parent game id. */
  game_id: number;
  /** Which team scored. */
  scoring_team: VolleyballScoringTeam;
  /** Notable-play tag on this point (006-002): 'Ace' | 'Block' | 'Spike' | 'Dive', or null when untagged. */
  event_type: VolleyballEventType | null;
  /** Server-assigned timestamp (DEFAULT now()). */
  recorded_at: string;
  /** Row creation time. */
  created_at: string;
}

/** Derived cumulative score for a game (COUNT of points per team). */
export interface VolleyballScore {
  /** Points scored by Team A ("SR"). */
  sr: number;
  /** Points scored by Team B (the opponent). */
  opponent: number;
}

/** A game enriched with its points and derived score. */
export interface VolleyballGameDetail {
  /** The game record. */
  game: VolleyballGame;
  /** The game's points in recorded order (recorded_at asc). */
  points: VolleyballPoint[];
  /** Score derived by counting points per team. */
  score: VolleyballScore;
}

/** A completed game with its final derived score (history list). */
export interface VolleyballHistoryGame {
  /** The game record. */
  game: VolleyballGame;
  /** Final score derived by counting points per team. */
  score: VolleyballScore;
}

/**
 * Response for GET /api/sports/volleyball/active.
 * `game` is null when no game is active.
 */
export interface VolleyballActiveResponse {
  /** The active game, or null when none exists. */
  game: VolleyballGameDetail | null;
}

/** Response for GET /api/sports/volleyball (game history). */
export interface VolleyballHistoryResponse {
  /** Completed games, sorted descending by completed_at. */
  games: VolleyballHistoryGame[];
}

/** Request body for POST /api/sports/volleyball (create a game). */
export interface VolleyballCreateGameRequest {
  /** Opponent team name (Team B). Team A is always "SR" and is not prompted. */
  team_b_name: string;
  /** SR side's partner jersey number (mandatory at creation; 006-002). */
  partner_number: number;
  /** SR side's partner name (optional; 006-002). */
  partner_name?: string | null;
}

/**
 * Response for POST /api/sports/volleyball when the single-active guard
 * refuses creation. HTTP 409; `blocked_by` describes the active game.
 */
export interface VolleyballBlockedResponse {
  /** Detail message explaining the refusal. */
  detail: string;
  /** The active game blocking creation. */
  blocked_by: VolleyballGame;
}

/** Request body for POST /api/sports/volleyball/{id}/points. */
export interface VolleyballAddPointRequest {
  /** Which team scored. */
  scoring_team: VolleyballScoringTeam;
  /** Optional notable-play tag written onto THIS point at creation
   *  (006-002; Bug T08-3: the UI holds a selected event and writes it with
   *  the next point). Null/omitted records an untagged point. */
  event_type?: VolleyballEventType | null;
}

/** Request body for POST /api/sports/volleyball/{id}/points/latest/event (006-002). */
export interface VolleyballTagEventRequest {
  /** The notable-play tag to write onto the most recently recorded point. */
  event_type: VolleyballEventType;
}

File: backend/api/volleyball.py
"""
Volleyball Scorekeeping API Endpoints
=====================================

FastAPI endpoints for the Beach Volleyball scorekeeping feature
(volleyball.games / volleyball.points).

Endpoints (contract from .features/designs_active/006-001_design.md):
    GET    /api/sports/volleyball                      -> game history (completed_at desc, w/ scores)
    GET    /api/sports/volleyball/active               -> active game + points + derived score, or game: null
    POST   /api/sports/volleyball                      -> create game (opponent name only); 409 if one is active
    POST   /api/sports/volleyball/{id}/points          -> add a point for a team
    DELETE /api/sports/volleyball/{id}/points/{team}   -> remove that team's most recent point
    POST   /api/sports/volleyball/{id}/end             -> end game (completed_at = MAX(recorded_at))
    DELETE /api/sports/volleyball/{id}                 -> abandon game (cascade deletes its points)
"""

from datetime import datetime

from fastapi import APIRouter, HTTPException

from backend_functions.queries import (
    get_game_detail,
    get_active_game,
    list_game_history,
    create_game,
    add_point,
    remove_last_point,
    end_game,
    abandon_game,
    VolleyballActiveGameExistsError,
    VolleyballNotFoundError,
    VolleyballStateError,
    VolleyballError,
)
from backend.schemas.volleyball_schemas import (
    VolleyballCreateGameRequest,
    VolleyballAddPointRequest,
)

router = APIRouter(prefix="/api/sports/volleyball", tags=["volleyball"])


def _serialize_game(game: dict) -> dict:
    """Normalize game datetime fields to ISO strings for JSON serialization."""
    if game is None:
        return None
    out = dict(game)
    for field in ("started_at", "completed_at", "created_at"):
        if isinstance(out.get(field), datetime):
            out[field] = out[field].isoformat()
    return out


def _serialize_point(point: dict) -> dict:
    """Normalize a point record's timestamps to ISO strings."""
    out = dict(point)
    for field in ("recorded_at", "created_at"):
        if isinstance(out.get(field), datetime):
            out[field] = out[field].isoformat()
    return out


def _serialize_detail(detail: dict) -> dict:
    """Shape a game detail (game + points + score) to the contract."""
    if detail is None:
        return None
    return {
        "game": _serialize_game(detail["game"]),
        "points": [_serialize_point(p) for p in detail["points"]],
        "score": detail["score"],
    }


@router.get("")
async def history():
    """Game history: completed games with final derived scores, completed_at desc."""
    try:
        return {"games": list_game_history()}
    except VolleyballError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to fetch volleyball history: {str(e)}")


@router.get("/active")
async def active():
    """Current active game with points and derived score, or game: null."""
    try:
        game = get_active_game()
        if game is None:
            return {"game": None}
        detail = get_game_detail(game["game_id"])
        return {"game": _serialize_detail(detail)}
    except VolleyballError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to fetch active volleyball game: {str(e)}")


@router.post("", status_code=201)
async def create(req: VolleyballCreateGameRequest):
    """Create a new game (opponent name only). 409 when one is already active."""
    try:
        game = create_game(team_b_name=req.team_b_name)
        return _serialize_game(game)
    except VolleyballActiveGameExistsError as e:
        blocking = _serialize_game(e.blocking_game)
        raise HTTPException(status_code=409, detail={"message": str(e), "blocked_by": blocking})
    except VolleyballError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to create volleyball game: {str(e)}")


@router.post("/{game_id}/points", status_code=201)
async def add(game_id: int, req: VolleyballAddPointRequest):
    """Record one point for a team on the active game."""
    try:
        point = add_point(game_id, scoring_team=req.scoring_team)
        return _serialize_point(point)
    except VolleyballNotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except VolleyballStateError as e:
        raise HTTPException(status_code=409, detail=str(e))
    except VolleyballError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to add point: {str(e)}")


@router.delete("/{game_id}/points/{scoring_team}")
async def remove(game_id: int, scoring_team: str):
    """Remove the most recent point recorded by ONE team (per-team undo)."""
    try:
        remove_last_point(game_id, scoring_team)
        return {"success": True}
    except VolleyballNotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except VolleyballStateError as e:
        raise HTTPException(status_code=409, detail=str(e))
    except VolleyballError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to remove point: {str(e)}")


@router.post("/{game_id}/end")
async def end(game_id: int):
    """End the game: status='completed', completed_at = MAX(recorded_at)."""
    try:
        game = end_game(game_id)
        return _serialize_game(game)
    except VolleyballNotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except VolleyballStateError as e:
        raise HTTPException(status_code=409, detail=str(e))
    except VolleyballError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to end volleyball game: {str(e)}")


@router.delete("/{game_id}")
async def abandon(game_id: int):
    """Abandon the game; its points are removed by FK cascade."""
    try:
        abandon_game(game_id)
        return {"success": True}
    except VolleyballNotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except VolleyballError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to abandon volleyball game: {str(e)}")
